common/solver/cg_kernels.cpp

```diff
@@ -54,85 +51,82 @@ namespace cg {
 
 
 template <typename ValueType>
-void initialize(std::shared_ptr<const OmpExecutor> exec,
+void initialize(std::shared_ptr<const DefaultExecutor> exec,
                 const matrix::Dense<ValueType> *b, matrix::Dense<ValueType> *r,
                 matrix::Dense<ValueType> *z, matrix::Dense<ValueType> *p,
                 matrix::Dense<ValueType> *q, matrix::Dense<ValueType> *prev_rho,
                 matrix::Dense<ValueType> *rho,
                 Array<stopping_status> *stop_status)
 {
-#pragma omp parallel for
-    for (size_type j = 0; j < b->get_size()[1]; ++j) {
-        rho->at(j) = zero<ValueType>();
-        prev_rho->at(j) = one<ValueType>();
-        stop_status->get_data()[j].reset();
-    }
-#pragma omp parallel for
-    for (size_type i = 0; i < b->get_size()[0]; ++i) {
-        for (size_type j = 0; j < b->get_size()[1]; ++j) {
-            r->at(i, j) = b->at(i, j);
-            z->at(i, j) = p->at(i, j) = q->at(i, j) = zero<ValueType>();
-        }
-    }
+    run_kernel(
+        exec,
+        [] GKO_KERNEL(auto row, auto col, auto b, auto r, auto z, auto p,
+                      auto q, auto prev_rho, auto rho, auto stop) {
+            if (row == 0) {
+                rho[col] = zero(rho[col]);
+                prev_rho[col] = one(prev_rho[col]);
+                stop[col].reset();
+            }
+            r(row, col) = b(row, col);
+            z(row, col) = p(row, col) = q(row, col) = zero(z(row, col));
```

Comment:
Unfortunately no, since on CUDA and HIP, we need `as_cuda/hip_type<ValueType>` instead

Reply:
Can we define a struct called `as_device_type` that does the correct thing based on which backend is being compiled? That could be used here, and it could also help unify your Cuda and hip simple kernel infrastructure.

Reply:
Yes, that makes sense. But I would still prefer `zero(z(row, col))` over `zero<as_device_type<ValueType>>()`

Reply:
I just thought it's un-intuitive to have a "zero of some vector", while the lambda is written on a scalar level. Otherwise it's fine.

Reply:
We now have `as_device_type` and `device_type` type aliases, but I guess it's still shorter to use an existing variable.